### ISSUE TITLE
Adding assert()-class methods for List/Dict verification via .asDict()/.asList()

### ISSUE THREAD

**Issue by egberts** [pyparsing/pyparsing#119]
Since I've been doing lots of unit tests associated with the pyparsing and Python List/Dict, there might be some benefit from creating a couple of `self.assertParserElementListTrue`-related functions to assist the pyparsing community with ensuring that nothing gets broken as we go along.

I've got a rough prototype of which I used heavily in my ISC Bind9/DHCP for such verification that List/Dict got constructed ... exactly and precisely:

```python
def assertParseElement(a_parse_element, a_test_data, a_expected_result,
                       a_assert_flag=True):
    """
    A nice unit test tool which provides an assert()-like function
    that takes an string, parse the string, takes its computed
    Pythonized list/dict and compares the result against its
    expected Pythonized result.
    :param a_parse_element:  ParserElement class to exercise
    :param a_test_data:  A string in which to be parsed by a_parse_element
    :param a_expected_result:  A Python list in which to expect
    :param a_assert_flag:  If True, then expected result must match or an
                           exception gets raised.
                           If False, then parse MUST fail or expected
                           result does not match, else an exception
                           gets raised
    :return: Always returns True (exception handles the False, like
             an assert() class would do)
    """
```
Is this something that our wonderful pyparsing community can use to ensure that such construct of List/Dict get done correctly?

https://github.com/egberts/pyparsing/blob/9a06cc2e5c47228db612a17fc68d6a931b4425db/test/test_isc_bind_aml.py#L72

**Comment by egberts** [comment]
| * need 2 expected arguments, one for the results as a list, one for results as a dict, default to None to indicate no test validation for list or dict

I see that you gravitated toward `assertParserResult...`, over my `assertParserElement`.  I chose the original class name 'ParserElement' because that was the core component being tested (as opposed to action-oriented)

Just mixing it up here to see which is more readable or better understood.
```
assertParseResultDictTrue()
assertParseResultDictFalse()
assertParseElementListTrue()
assertParseElementListFalse()
```

**Comment by egberts** [comment]
```python
assertParserResultDictTrue(
    parser_element,
    test_string,
    expected_dict_result,
    optional_message
)
```
I can go with that.

-- Still struggling on this context design of yours.  This pertains primarily to `self.`?

**Comment by egberts** [comment]
or is it:
```python
    @contextlib.contextmanager
    def sub.Test(self, msg='', **params):
```

**Comment by egberts** [comment]
Also, do we want separate `test_strings` and `expected_results` arguments?

Or do we want to tie the two closer together and call it `test_data` like the following pair-list example?
```python
test_csv_data = {
    [
        {'abc, def, ghi', 'my_csv_row': ['abc', 'def', 'ghi']},
        {'123, 345, 567', 'my_csv_row': ['123', '345', '567']}
    ]
} 
myCSVParserElement.assertDictEqual(test_csv_data, 'it did not parse.')
```

**Comment by egberts** [comment]
f've completed my ISC Bind9 project.  Sent you a sourceforge email with the details.

**Comment by ptmcg** [comment]
Here's what I have so far, after trying it out on a few of the asserts in unitTests.py:

```
class pyparsing_test:
    class ParseTestContext:
        def __init__(self, test_case, expr):
            self._test_case = test_case
            self._expr = expr
            self._save_context = {}

        def __enter__(self):
            self._save_context['default_whitespace'] = ParserElement.DEFAULT_WHITE_CHARS
            self._save_context['default_keyword_chars'] = Keyword.DEFAULT_KEYWORD_CHARS
            self._save_context['literal_string_class'] = ParserElement._literalStringClass
            self._save_context['packrat_enabled'] = ParserElement._packratEnabled
            return self

        def __exit__(self, *args):
            # reset pyparsing global state
            if ParserElement.DEFAULT_WHITE_CHARS != self._save_context['default_whitespace']:
                ParserElement.setDefaultWhitespaceChars(self._save_context['default_whitespace'])
            Keyword.DEFAULT_KEYWORD_CHARS = self._save_context['default_keyword_chars']
            ParserElement.inlineLiteralsUsing(self._save_context['literal_string_class'])
            for diagname in __diag__._all_names:
                __diag__.disable(diagname)
            ParserElement._packratEnabled = self._save_context['packrat_enabled']

        def assertParseResultEquals(self, test_string, expected_list=None,expected_dict=None,
                                    parseAll=True, msg=None, verbose=True):
            message = "{} failed parsing {!r}".format(self._expr, test_string)
            if msg is not None:
                message = "{}: {}".format(msg, message)

            result = self._expr.parseString(test_string, parseAll=parseAll)
            if verbose:
                print(result.dump())
            if expected_list is not None:
                self._test_case.assertEqual(expected_list, result.asList(), msg=message)
            if expected_dict is not None:
                self._test_case.assertEqual(expected_dict, result.asDict(), msg=message)

        @contextmanager
        def assertRaises(self, test_string, exc_type=ParseException, parseAll=True, msg=None):
            message = "{} failed to raise exception while parsing {!r}".format(self._expr, test_string)
            if msg is not None:
                message = "{}: {}".format(msg, message)

            with self._test_case.assertRaises(exc_type, msg=message):
                self._expr.parseString(test_string, parseAll=True)

```
And a sample changed test:

```
        alphas = ppu.Greek.alphas
        greet = pp.Word(alphas) + ',' + pp.Word(alphas) + '!'

        # input string
        hello = "Καλημέρα, κόσμε!"
        result = greet.parseString(hello)
        print(result)
        self.assertTrue(result.asList() == ['Καλημέρα', ',', 'κόσμε', '!'],
                        "Failed to parse Greek 'Hello, World!' using "
                        "pyparsing_unicode.Greek.alphas")
```
becomes:

```
        alphas = ppu.Greek.alphas
        greet = pp.Word(alphas) + ',' + pp.Word(alphas) + '!'

        with ppt.ParseTestContext(self, greet) as ctx:
            # input string
            hello = "Καλημέρα, κόσμε!"

            ctx.assertParseResultEquals(hello,
                                        expected_list=['Καλημέρα', ',', 'κόσμε', '!'],
                                        msg="Failed to parse Greek 'Hello, World!' using "
                                            "pyparsing_unicode.Greek.alphas",
                                        verbose=True)
```

The context manager part really comes into play with tests that modify some of the pyparsing settings like packrat parsing or default whitespace characters.

Still need to make up my mind on whether the expr should be part of the context. The current design puts the expr definition outside of the `with` statement, and I'm not sure I like that. The alternative would look something like:

```
        with ppt.ParseTestContext(self) as ctx:
            alphas = ppu.Greek.alphas
            greet = pp.Word(alphas) + ',' + pp.Word(alphas) + '!'

            # input string
            hello = "Καλημέρα, κόσμε!"

            result = greet.parseString(hello, parseAll=True)
            ctx.assertParseResultEquals(result,
                                        expected_list=['Καλημέρα', ',', 'κόσμε', '!'],
                                        msg="Failed to parse Greek 'Hello, World!' using "
                                            "pyparsing_unicode.Greek.alphas",
                                        verbose=True)
```
I'm thinking this may read better.

**Comment by ptmcg** [comment]
Now thinking defining a mixin class would be better, and breaking out the contextmanager part as a separate class (explicit better than implicit):

```
class pyparsing_test:
    class ParseTestContext:
        def __init__(self):
            self._save_context = {}

        def __enter__(self):
            self._save_context['default_whitespace'] = ParserElement.DEFAULT_WHITE_CHARS
            self._save_context['default_keyword_chars'] = Keyword.DEFAULT_KEYWORD_CHARS
            self._save_context['literal_string_class'] = ParserElement._literalStringClass
            self._save_context['packrat_enabled'] = ParserElement._packratEnabled
            return self

        def __exit__(self, *args):
            # reset pyparsing global state
            if ParserElement.DEFAULT_WHITE_CHARS != self._save_context['default_whitespace']:
                ParserElement.setDefaultWhitespaceChars(self._save_context['default_whitespace'])
            Keyword.DEFAULT_KEYWORD_CHARS = self._save_context['default_keyword_chars']
            ParserElement.inlineLiteralsUsing(self._save_context['literal_string_class'])
            for diagname in __diag__._all_names:
                __diag__.disable(diagname)
            ParserElement._packratEnabled = self._save_context['packrat_enabled']


    class ParseResultsAsserts(unittest.TestCase):
        """
        mixin class to add ParseResults assert method to TestCase classes

            class MyTestCase(unittest.TestCase, ppt.ParseResultsAsserts):
                ...
        """
        def assertParseResultEquals(self, result, expected_list=None, expected_dict=None, msg=None):
            if expected_list is not None:
                self.assertEqual(expected_list, result.asList(), msg=msg)
            if expected_dict is not None:
                self.assertEqual(expected_dict, result.asDict(), msg=msg)

        @contextmanager
        def assertRaises(self, exc_type=ParseException, msg=None):
            with self.assertRaises(exc_type, msg=msg):
                yield
```

**Comment by egberts** [comment]
As a recap, the current but remaining generalized problems related to ParseElement-related unittest approaches that I am (and possibly you are as well) facing remains:

1. Syntax verification (via assert)
2. asList() verification
3. asDict() verification

Jack of all trade assert or separate focus?  

At the moment, looks like a separate class/method looks cleaner and makes for a smaller main package.

**Comment by egberts** [comment]
Example code

== By-Argument ==
```python
assertParseResults(pe, string, expected_failure=False, expected_list=[‘value’], expected_dict={})
```
== By-Method ==
```python
assertParseResultsDictEqual(pe, string, {‘value’})
assertParseResultsListEqual(pe, string, [‘value])
assertParseResultsValid(pe, string)
```

I do do see a need to provide a list of string for multiple tests such as .runTest() excellent so provided but have not figure out a way to represent those to asDict()/asList() tests.

**Comment by klahnakoski** [comment]
I think the runTest() method already exists in the codebase, albeit hidden:

https://github.com/pyparsing/pyparsing/blob/87016995f05a0b4927eef22f767a349d85e6afd2/simple_unit_tests.py#L28

It exists as a method that scans for test structures, like you propose, and then compares them to the expected result.

When the comparison is pulled out, we get 

https://github.com/klahnakoski/pyparsing/blob/5653dbdadedd9c4667f9b40d61990c3703eef5d3/test/test_simple_unit.py#L25

...of course all the structures are turned into runTest() calls:

https://github.com/klahnakoski/pyparsing/commit/bcaf36ce89f2bdf36817517a7eff610250d5170c?w=1

Please look over the `test` dir in my `remove-dict` branch.  The mess I have made in that branch is not intended for public consumption, but the `test` directory has a cleaned-up test suite.  I can provide a a clean PR to pyparsing if you wish.

https://github.com/klahnakoski/pyparsing/tree/remove-dict/test

**Comment by ptmcg** [comment]
I am set dead against splitting `assertParseResultsEquals` into a dict version and a list version. Why should there be two calls to verify a ParseResults content? Even worse, to have a call that embeds the parseString call. I had that in my original version, but it broke down in practice because there are many variations where a ParseResults may come from, so it is cleaner to put the parseString (or whatever ParseResults-generating code) back explicitly in the test, and have the assert method just verify the results.

Any code or convenience classes in simple_unit_tests.py are intended just for that module, with the intent that the tests themselves could be defined just by using some static namedtuples. I suspect these new assert methods will get retrofitted into that module too (not the other way around).

I will split off a `unittest_integration` branch, and drop my current work in there. I've converted a number of tests in unitTests.py to using them, and I think they generally look decent. There are some attempts at integrating with `runTests` that could still stand some improvement.

**Comment by egberts** [comment]
My bad.  Perhaps I was looking for convenience and not on avoiding duplicitous CPU effort.

**Comment by ptmcg** [comment]
No I get it. I'm all for convenience - if tests are hard to write, they are less likely to get written. But I don't agree that two separate calls is more convenient than one complex call. To my mind it is kind of like having an assert API for Python lists, with separate `assertListLen` and `assertListContents` methods, or a dict API with separate `assertDictKeys` and `assertDictValues` methods. In the case of ParseResults, we have a single thing with potentially several characteristics to be validated, but I feel the singleness of the thing corresponds nicely with the singleness of the assert method.

So for convenience' sake, I could dispense with the named args 'expected_list' and 'expected_dict' completely, and just crack an unnamed `*args` for a list and/or dict, similar to what is done in the runTests results asserter. Just passing in a list and/or dict to `assertParseResultsEquals` would be convenient. I already did this by accident in a couple of places when I just had an expected list, and the arg ordering just (luckily) did the right thing. I don't *really* even need the named arg for the message, but this technique may be best used sparingly. Or pass `expected` in as a list, a dict or a (list, dict) tuple, and then infer the test's assert requirements based on the type of value passed.

I was also thinking at one time along the lines of a context manager for some retention of parse expressions across tests, but in the end, its most useful purpose is for resetting global state. Please look at the latest commit to the `unittest_integration` branch  It turns out that the state resetting just ended up in the base class `_runTest` method, which I wasn't expecting, but now seems obvious.

I also took a stab at verifying the results from `runTests`, but I've never liked trying to keep tests and expected results in separate lists, to be rezipped later. Still I'm kind of stuck with `runTests` API. I suppose I could look at modifying `runTests` to accept some kind of optional embedded "#EXPECTED" value after each test, similar to the leading comment handling. That wouldn't necessarily break existing code using runTests. and would keep the expected results close to the input test string.

Thanks for weighing in on this - it was your original suggestion and I'm glad you made it. Looking at my code coverage (currently about 87%) I've made a list of a number of missing unit tests based on the uncovered code. Having these methods will definitely help that work.

**Comment by ptmcg** [comment]
And after writing all that, I'm thinking that maybe there is an 80% case where we just want to specify a parser, an input string to be parsed using parseString, and an expected list of parsed tokens. That's pretty much your `assertParseResultsList` method. Would we really need True and False versions of this? Would be a nice check to see where in unitTests.py we could replace 5 or 6 lines of code with just this one call. I would bet there are at least a dozen.

**Comment by ptmcg** [comment]
Looking at your proposed methods now in https://github.com/pyparsing/pyparsing/issues/119#issuecomment-525758396. In contrast to `assertParseResultsEquals`. The signatures make sense - each takes a parser, an input string, and either an expected list or expected dict, to be checked after calling `parser.parseString(input_string)`. Can we come up with a name that conveys that we aren't just validating a ParseResults that was somehow created and passed in, but is also actually calling parseString, without it being 400 characters long.  `assert"call_parseString_on_parser_with_input_string_and_validate_tokens"List` and the same for Dict. Maybe something like `assertParseAndCheckList/Dict`? Again, I think "Equals" or "True" are probably already understood, and I'm having a hard time coming up with a "False'" or "NotEquals" flavor.

Sorry for being slow to come around on this, I like thinking through design alternatives for a bit. After rereading your notes, I started envisioning you writing and rewriting these methods as wrappers around the `assertParseResultsEquals` in subsequent projects, and they might just as well be in the base lib.

**Comment by klahnakoski** [comment]
@ptmcg  Please make the [`pyparsing_test`](https://github.com/pyparsing/pyparsing/commit/76d9748718eb3fbf16b2e352fd524ae123dd9f47?w=1#diff-ea9b784481a02f9eac5b2944fab65a9eR6708) class a module in the `test` directory; modules are better namespace containers; plus the code should be migrating away from one-giant-file.

Also, if you make a PR that points to your branch, then code discussions can happen there; whether or not it gets merged.

**Comment by ptmcg** [comment]
Ok, well, wherever they end up going, here is how the two convenience assert methods end up looking:

        def assertParseAndCheckList(self, expr, test_string, expected_list, msg=None):
            result = expr.parseString(test_string, parseAll=True)
            self.assertParseResultsEquals(result, expected_list=expected_list, msg=msg)

        def assertParseAndCheckDict(self, expr, test_string, expected_dict, msg=None):
            result = expr.parseString(test_string, parseAll=True)
            self.assertParseResultsEquals(result, expected_dict=expected_dict, msg=msg)

**Comment by ptmcg** [comment]
No further feedback or comments in almost 3 weeks, so I'm going to check these in as-is. Then let's address the code reorg as a separate topic.

**Comment by ptmcg** [comment]
These suggestions have been incorporated into the testing refactor work to be released in 3.0.0, thanks!
